core/core.cc

```diff
@@ -434,10 +466,62 @@ App::App(int argc, char **argv)
       graph.kernel.iterations = value;
     }
 
-    if (!strcmp(argv[i], "-and")) {
+    if (!strcmp(argv[i], "-and")) { //add a new task graph
       graphs.push_back(graph);
       graph = default_graph();
     }
+
+    // -- add by Yuankun
+    if (!strcmp(argv[i], "-num_src_input")) {
+      needs_argument(i, argc, "-num_src_input");
+      long value  = atol(argv[++i]);
+      if (value < 0) {
+        fprintf(stderr, "error: Invalid flag \"-num_src_input %ld\" must be >= 0\n", value);
+        abort();
+      }
+      graph.kernel.kernel_arg.num_src_input = value;
+      graph.kernel.kernel_arg.input_data = (char **)malloc(sizeof(char*) * value);
+      graph.kernel.kernel_arg.input_bytes_per_src = (size_t *)malloc(sizeof(size_t) * value);
+    }
```

Comment:
Why do you need user to specify how many input data does a kernel take? Doesn't it should be associated with kernel type? Once a kernel is specified, then this number is automatically set.

Reply:
I could remove this kernel arg parameter and set it automatically.

Reply:
If we need scratch space per task, I think we should specify that separately. We should not be trying to use task input or output to do that, because that will also cause additional communication to occur.

Reply:
Should we provide the space in core or let runtime to provide? If we do that in core, maybe implement a memory pool? but it requires locks, which may hurt the performance.